### PR TITLE
[2565] - Geocoder setup

### PR DESCRIPTION
**Pull request by TobyRet**
### Context

This PR adds geocoding to the Site and Provider models. When either of these models is created or updated, a background job is triggered to search and store the latititude and longitude for each of their addresses. Note the background jobs are only triggered if address details have changed.

### Changes proposed in this pull request
- Stubbing of Geocoder requests
- Creation of geocoding background job
- Triggering of geocoding background job in the Site and Provider model
- Creation of Geocoder config (uses Google for search, adds caching)

### Guidance to review
- Aiming for no more than three rejections :-)

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by TobyRet** [comment]
[Add geocoding setup](https://trello.com/c/hDXI6Vse/2565-add-geocoding-setup)

File: app/jobs/geocode_job.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+class GeocodeJob < ApplicationJob
```
Comment (dankmitchell) [inline app/jobs/geocode_job.rb]: Nice!

File: spec/support/helpers.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
+module Helpers
```
Comment (defong) [inline spec/support/helpers.rb]: Can we have that by itself
`helper` just seems too generic and probably make this file grow to monster

image

File: config/settings/development.yml [hunk 3]
```diff
@@ -4,3 +4,5 @@ mcbg:
   redis_password:
 system_authentication_token: "Ge32"
 log_level: debug
+google:
```
Comment (dankmitchell) [inline config/settings/development.yml]:
```suggestion
gcp:
```

File: config/settings/pentest.yml [hunk 4]
```diff
@@ -6,3 +6,5 @@ bg_jobs:
     cron: "0 */6 * * *"
     class: "BulkSyncCoursesToFindJob"
     queue: find_sync
+google:
```
Comment (dankmitchell) [inline config/settings/pentest.yml]:
```suggestion
gcp:
```

**Comment by defong** [comment]
looks like it is suffering from the same fate as #1014 @misaka  @timabell 

image

File: spec/support/helpers.rb (hunk 2 above)
Comment (TobyRet) [reply spec/support/helpers.rb]: Yep good idea

**Comment by TobyRet** [comment]
Screenshot 2019-11-29 at 09 57 36

Boo. So I realised that I was only geocoding after updating the address in a Provider and Site. We also want to do this when creating Provider and Site but this is causing stack overflow errors

File: app/models/provider.rb [hunk 5]
```diff
@@ -120,6 +120,17 @@ def current_accredited_courses
...
   before_discard { discard_courses }
 
+  geocoded_by :full_address
+  after_commit -> { GeocodeJob.perform_later("Provider", id) }
+
+  def full_address
+    [address1, address2, address3, address4, postcode].compact.join(", ")
+  end
+
+  def address_changed?
```
Comment (dankmitchell) [inline app/models/provider.rb]: Is this method being used anymore?

Comment (TobyRet) [reply app/models/provider.rb]: Yep it's back in there now

**Pull request merged**